add uniform processors for altclip + chinese_clip

molbap opened the pull request:
# What does this PR do?

Adds two models for https://github.com/huggingface/transformers/pull/30511 , see parent PR for more details

molbap committed:
add initial design for uniform processors + align model

molbap committed:
add uniform processors for altclip + chinese_clip

molbap commented:
Tests will succeed when #31197 is merged. See #30511 tests to confirm.

molbap committed:
fix mutable default :eyes:

molbap committed:
add configuration test

molbap committed:
handle structured kwargs w defaults + add test

molbap committed:
protect torch-specific test

molbap committed:
fix style

molbap committed:
fix

molbap committed:
rebase

molbap committed:
update processor to generic kwargs + test

molbap committed:
fix style

molbap committed:
add sensible kwargs merge

molbap committed:
update test

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31198). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts reviewed (commented):
Thanks for working on this! 

Overall looks good to me. Main thought is that we probably want to move a lot of the kwarg preparation logic outside of the main `__call__` method, and it can probably be abstracted into something more general for most processors. 

Regarding tests, it would be good to check defaults and updates are correctly passed for all of the possible kwargs, particularly defaulting to the tokenizer.init_kwargs and ensuring these can be overridden. 

Would be good to get a second opinion from @qubvel here!

molbap commented:
Thanks! I'm currently moving the kwargs merging in a common method actually, and explaining order of operations wrt to kwargs priorities.

molbap committed:
fix assertEqual

molbap committed:
move kwargs merging to processing common

qubvel reviewed (commented):
Great work 🤗  Very structured and clean code!
I have the same concern as @amyeroberts that the logic of merging is worth split/move to separate methods. It is a bit hard to follow merging steps of different kwargs.

What do you think about grouping merging strategy per modality instead of per step?
not sure if its possible, because some kwargs, probably, have to be popped in advance.

Something like this:
```python
images_kwargs = {} if images_kwargs is None else images_kwargs
default_images_kwargs = ChineseClipProcessorKwargs._defaults.get("images_kwargs", {}).copy()

# merging vision kwargs
images_kwargs = {**default_images_kwargs, **images_kwargs}   # with default
images_kwargs = {**images_kwargs, **kwargs.pop("images_kwargs", {})}  # with passed dict
images_kwargs = merge_with_kwargs_by_key(images_kwargs, ...)
images_kwargs = merge_with_common(images_kwargs, common_kwargs)
```

molbap commented on src/transformers/processing_utils.py:
```diff
@@ -54,6 +68,190 @@
+    do_normalize: Optional[bool]
+    image_mean: Optional[Union[float, List[float]]]
+    image_std: Optional[Union[float, List[float]]]
+    do_pad: Optional[bool]
```
sure! adding it

molbap replied:
for your comment above, kwargs merging is broken down in https://github.com/huggingface/transformers/pull/31197 in a common method, that's a bit more up-to-date (apologies for the split PRs)

molbap commented on tests/models/altclip/test_processor_altclip.py:
Moving it to a common testing instance!

amyeroberts replied:
I think this needs to be moved to under the altclip model folder

molbap replied:
I'm not sure I understand why? Is it because there was no test before?

amyeroberts replied:
Ah, I think I misunderstood your previous comment: I thought you meant move this file under a higher level e.g. `tests/test_processor_xxx.py` or something

molbap committed:
rework kwargs for type hinting

molbap committed:
just get Unpack from extensions

molbap committed:
run-slow[align]

molbap committed:
handle kwargs passed as nested dict

molbap committed:
add from_pretrained test for nested kwargs handling

molbap committed:
[run-slow]align

molbap committed:
update documentation + imports

molbap committed:
update audio inputs

molbap committed:
protect audio types, silly

molbap committed:
try removing imports

molbap committed:
make things simpler

molbap committed:
simplerer

molbap committed:
move out kwargs test to common mixin

molbap committed:
[run-slow]align

molbap committed:
skip tests for old processors

molbap committed:
[run-slow]align, clip

molbap committed:
!$#@!! protect imports, darn it

molbap committed:
[run-slow]align, clip

molbap committed:
Merge branch 'main' into uniform_processors_1

molbap committed:
[run-slow]align, clip

molbap committed:
fix conflicts

molbap committed:
update common processor testing

molbap committed:
add altclip

molbap committed:
add chinese_clip

molbap committed:
add pad_size

molbap committed:
[run-slow]align, clip, chinese_clip, altclip

molbap committed:
remove duplicated tests

molbap commented:
I simplified the addition/testing of uniformized processors on that one. It's pulling changes from #31197 so the file diff will be reduced once that first one is merged! shoud be ready for review after that.

molbap committed:
fix

molbap commented:
This one should be ready for review @amyeroberts after #31197 is reviewed&merged!

molbap committed:
update doc

molbap committed:
improve documentation for default values

molbap committed:
add model_max_length testing

This parameter depends on tokenizers received.

molbap committed:
Raise if kwargs are specified in two places

molbap committed:
fix

molbap committed:
Merge branch 'uniform_processors_1' into uniform_processors_2

molbap committed:
Merge branch 'main' into uniform_processors_2

molbap committed:
Merge branch 'main' into uniform_processors_2

molbap committed:
match defaults

molbap committed:
force padding

molbap committed:
fix tokenizer test

molbap committed:
clean defaults

molbap committed:
move tests to common

amyeroberts commented on src/transformers/models/chinese_clip/processing_chinese_clip.py:
```diff
@@ -17,9 +17,25 @@
-from ...processing_utils import ProcessorMixin
-from ...tokenization_utils_base import BatchEncoding
+from ...image_utils import ImageInput
+from ...processing_utils import (
+    ProcessingKwargs,
+    ProcessorMixin,
+)
```
nit 
```suggestion
from ...processing_utils import ProcessingKwargs, ProcessorMixin
```

amyeroberts commented on tests/models/altclip/test_processor_altclip.py:
```diff
@@ -0,0 +1,70 @@
+import unittest
```
Missing copyright header

amyeroberts commented on tests/models/altclip/test_processor_altclip.py:
```diff
@@ -0,0 +1,70 @@
+    def get_image_processor(self, **kwargs):
+        return CLIPImageProcessor.from_pretrained(self.model_id, **kwargs)
+
+    def tearDown(self):
+        pass
+
```
What's the reason for the pass here? I'm guessing the parent tearDown is removing objects which aren't present for this tester?

molbap replied:
initially, yes, but now it's just not used :grimacing: will remove

amyeroberts commented on tests/test_processing_common.py:
```diff
@@ -126,13 +129,12 @@ def test_tokenizer_defaults_preserved_by_kwargs(self):
             self.skipTest(f"image_processor attribute not present in {self.processor_class}")
         image_processor = self.get_component("image_processor")
-        tokenizer = self.get_component("tokenizer", max_length=117)
+        tokenizer = self.get_component("tokenizer", max_length=117, padding="max_length")
```
Does this mean the padding behaviour of the processors isn't backwards compatible i.e. it used to default to `"max_length"` and doesn't anymore? Or is this because of the new processors being tested?

molbap replied:
the latter, this is just because of new processors being tested - no weird BC behaviour here as far as I know

amyeroberts commented on tests/test_processing_common.py:
```diff
@@ -60,7 +60,10 @@ def get_component(self, attribute, **kwargs):
 
         component_class = processor_class_from_name(component_class_name)
-        component = component_class.from_pretrained(self.tmpdirname, **kwargs)  # noqa
+        try:
+            component = component_class.from_pretrained(self.tmpdirname, **kwargs)  # noqa
+        except AttributeError:
+            component = component_class.from_pretrained(self.model_id, **kwargs)  # noqa
```
Why do we need the try/except patten here?

molbap replied:
because some processors are initialized with a self.tmpdirname, some don't have this attribute but have a self.model_id  instead - I can check for attribute presence/absence instead of try/except, it'd be cleaner

amyeroberts reviewed (commented):
Thanks for working on this! 

Just some small comments / questions

molbap commented on src/transformers/models/altclip/processing_altclip.py:
```diff
@@ -35,22 +58,21 @@ class AltCLIPProcessor(ProcessorMixin):
-    def __init__(self, image_processor=None, tokenizer=None, **kwargs):
-        feature_extractor = None
-        if "feature_extractor" in kwargs:
+    def __init__(self, image_processor=None, tokenizer=None, feature_extractor=None):
```
ah true, will try that, should do nicely!

molbap replied:
done!

molbap committed:
remove try/catch block

molbap committed:
deprecate kwarg

molbap committed:
format

molbap committed:
add copyright + remove unused method

molbap committed:
Merge branch 'main' into uniform_processors_2

molbap committed:
[run-slow]altclip, chinese_clip

molbap commented:
@amyeroberts I forgot to launch slow tests, they are a bit slow, but should be good for review now!
- removed try/except block for processors attributes
- formatting
- removed unused teardown method

amyeroberts commented on src/transformers/models/altclip/processing_altclip.py:
```diff
@@ -35,22 +58,16 @@ class AltCLIPProcessor(ProcessorMixin):
             The image processor is a required input.
         tokenizer ([`XLMRobertaTokenizerFast`], *optional*):
             The tokenizer is a required input.
+        feature_extractor ([`CLIPFeatureExtractor`], *optional*):
+            The feature extractor is a deprecated input.
```
Given we don't advertise `audio` is an input when calling the processor, do we want to add `feature_extractor` here to the docstring?

molbap replied:
Was hesitant on that - wdyt about not advertising unused modalities, rather? I wanted to have the same signature everywhere but that might be overkill

molbap replied:
oh, no I understand the issue here @amyeroberts , in this case `feature_extractor` is unfortunately a backwards compatible name for the image processor. So we cannot remove it.

amyeroberts replied:
Yes, I think we need to be able to accept feature_extractors as an input, but I don't think they should be added to the docstring as it's confusing. This probably falls under the same scope for @qubvel's deprecation decorators again

molbap replied:
agreed! done!

amyeroberts commented on src/transformers/models/altclip/processing_altclip.py:
```diff
@@ -16,10 +16,33 @@
+from ...utils.deprecation import deprecate_kwarg
+
+
+# TODO (@molbap) This is a bother, forward references from TypedDict are resolved and need this to work
```
Do we have an issue open for this, so that we don't forget?

molbap replied:
ah, this is outdated - with the new import structure of Unpack this issue is not there anymore!

amyeroberts commented on src/transformers/models/altclip/processing_altclip.py:
```diff
@@ -95,13 +117,24 @@ def __call__(self, text=None, images=None, return_tensors=None, **kwargs):
+            raise ValueError("You must specify either text or images.")
+        output_kwargs = self._merge_kwargs(
+            AltClipProcessorKwargs,
+            tokenizer_init_kwargs=self.tokenizer.init_kwargs,
```
Do you think we should have an equivalent `init_kwargs` for the image processors and feature extractors for consistency? Or better to leave just in one place?

molbap replied:
currently the image processors and other extractors don't need this to work (this is mostly for fast tokenizers), not opposed to including it as well but it seems to be more a tokenizers thing?

amyeroberts replied:
OK, let's leave for now!

amyeroberts commented on src/transformers/models/altclip/processing_altclip.py:
```diff
@@ -16,10 +16,33 @@
-from ...processing_utils import ProcessorMixin
-from ...tokenization_utils_base import BatchEncoding
+
+try:
+    from typing import Unpack
+except ImportError:
+    from typing_extensions import Unpack
```
Rather than try/except could we do this in a more explicity way e.g. by testing on the version in the env?

molbap replied:
done!

amyeroberts reviewed (commented):
Looks great - just a few small things here and there! 

cc @ArthurZucker for reference on the changes in processors and their kwargs

molbap committed:
clean imports

molbap committed:
fix version

qubvel commented on src/transformers/models/altclip/processing_altclip.py:
```diff
@@ -35,22 +58,16 @@ class AltCLIPProcessor(ProcessorMixin):
-            feature_extractor = kwargs.pop("feature_extractor")
-
+    @deprecate_kwarg(old_name="feature_extractor", version="5.0.0", new_name="image_processor")
+    def __init__(self, image_processor=None, tokenizer=None, feature_extractor=None):
```
As soon as we added `@deprecate_kwarg(old_name="feature_extractor", version="5.0.0", new_name="image_processor")` we no longer need `feature_extractor` in the function signature, an input keyword argument `feature_extractor` will be renamed to `image_processor`.

```suggestion
    def __init__(self, image_processor=None, tokenizer=None):
```

molbap replied:
yes indeed, thanks. should be good to go!

molbap committed:
clean up deprecation

molbap committed:
fix style

molbap commented:
@amyeroberts I think this one is ok for a review!

molbap committed:
add corner case test on kwarg overlap

amyeroberts commented on src/transformers/models/altclip/processing_altclip.py:
```diff
@@ -16,10 +16,23 @@
-from ...processing_utils import ProcessorMixin
-from ...tokenization_utils_base import BatchEncoding
+
+if sys.version_info >= (3, 11):
+    from typing import Unpack
+else:
+    from typing_extensions import Unpack
```
Similar to what we have with `PILImageResampling` could we define an object, where we do this importing once in one place and then import from there?

molbap replied:
ah that's a good point! will do that

amyeroberts reviewed (approved):
Nice! Thanks for all the work on this 🤗

molbap committed:
resume processing - add Unpack as importable

molbap committed:
Merge branch 'main' into uniform_processors_2

molbap commented:
Re-pinging @amyeroberts on that one and @qubvel and @zucchini-nlp optionally if you want to take another look - not much changed, just was between sprints and didn't merge that one :grimacing: should be good to go now (I know you approved, Amy, but that was a while ago!)

molbap committed:
Merge branch 'main' into uniform_processors_2

molbap committed:
add tmpdirname

molbap committed:
fix altclip

molbap committed:
fix up

molbap commented:
Tests that were working are broken since passing from `crop_size` to `size` in common tests I believe  - will look into this soon

molbap committed:
Merge branch 'main' into uniform_processors_2

molbap committed:
add back crop_size to specific tests

molbap committed:
Merge branch 'main' into uniform_processors_2

molbap committed:
generalize tests to possible video_processor

molbap committed:
add back crop_size arg

molbap committed:
fixup overlapping kwargs test for qformer_tokenizer

molbap committed:
remove copied from

molbap committed:
fixup chinese_clip tests values

molbap committed:
fixup tests - qformer tokenizers

molbap committed:
[run-slow] altclip, chinese_clip

amyeroberts commented on tests/models/altclip/test_processor_altclip.py:
```diff
@@ -0,0 +1,178 @@
+    def get_image_processor(self, **kwargs):
+        return CLIPImageProcessor.from_pretrained(self.model_id, **kwargs)
+
+    def prepare_image_inputs(self):
+        """This function prepares a list of PIL images, or a list of numpy arrays if one specifies numpify=True,
+        or a list of PyTorch tensors if one specifies torchify=True.
+        """
+
+        image_inputs = [np.random.randint(255, size=(3, 30, 400), dtype=np.uint8)]
+        image_inputs = [Image.fromarray(np.moveaxis(x, 0, -1)) for x in image_inputs]
+        return image_inputs
```
No need to add this now -- it's in athe processormixin

molbap replied:
right, done!

amyeroberts reviewed (approved):
Thanks for adding!

molbap committed:
remove prepare_image_inputs
